### PR TITLE
Ice emulator and ice-ocean-air balance

### PR DESCRIPTION
**Pull request by guillaumevernieres**
I removed the example that was in this repo and replaced it with a FFNN that will allow emulating ice-ocean-atmos balances. The unbalanced ssh project was a bit too compute intensive for my laptop and as a first project.

### Work done:
- This PR adds a simple yet realistic ice concentration emulator and makes use of the automatic differentiation tools of `Torch` to get the Jacobian.  
- The build system was changed and assumes this repo is built as part of a `jedi-bundle` such as the one found in the `GDASApp`.
- I removed the ctest for now, it will be added in a future PR, with the addition of a sub-sampled `.cdl` file ... because no git-lfs!

### Preliminary results
The FFNN that was tested is quite simple, only 19 parameters in total, trained on ~ 130,000 patterns/targets pairs. It generalizes well over multiple days and is small enough to be trained at every cycles.
Here's a figure of the true and predicted ice concentration along with the Jacobian elements:
pred-f003

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -2,22 +2,28 @@
 # DAML Applications                                                         #
 # ------------------------------------------------------------------------- #
 
-# Check for minimim cmake requirement
-cmake_minimum_required( VERSION 3.20 FATAL_ERROR )
+project(gdas-utils LANGUAGES C CXX )
```
Comment (CoryMartin-NOAA) [inline CMakeLists.txt]: not the right name?

File: CMakeLists.txt [hunk 2]
```diff
@@ -2,22 +2,28 @@
...
+set(CMAKE_C_STANDARD 11)
+set(CMAKE_C_STANDARD_REQUIRED ON)
+set(CMAKE_C_EXTENSIONS OFF)
+set(CMAKE_FORTRAN_STANDARD 08)
+set(CMAKE_FORTRAN_STANDARD_REQUIRED ON)
+set(CMAKE_FORTRAN_EXTENSIONS OFF)
 
-include(GNUInstallDirs)
-enable_testing()
+set(Torch_ROOT /home/gvernier/sandboxes/AI/pytorch/install-dec14)
```
Comment (CoryMartin-NOAA) [inline CMakeLists.txt]: Bad!

File: CPPLINT.cfg [hunk 3]
```diff
@@ -0,0 +1,3 @@
+set noparent
```
Comment (CoryMartin-NOAA) [inline CPPLINT.cfg:1]: :-(

File: IceEmul/IceEmul.h [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+    for (const auto& parameter : model_->parameters()) {
+        total_params += parameter.numel();
+    }
+    oops::Log::info() << "Number of parameters: " << total_params << std::endl;
+  }
+
+  // Training
+  void train(const torch::Tensor input, const torch::Tensor target) {
+    // Loss function and optimizer.
+    oops::Log::info() << "Define loss fction and optimizer " << std::endl;
```
Comment (CoryMartin-NOAA) [inline IceEmul/IceEmul.h]:
```suggestion
    oops::Log::info() << "Define loss function and optimizer " << std::endl;
```

File: IceEmul/IceEmul.h [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+      optimizer.step();
+    }
+    std::cout << std::endl;
+  }
+
+  // Prepare the inputs/targets
+  std::tuple<torch::Tensor, torch::Tensor,
+             std::vector<float>, std::vector<float>> prepData(std::string fileName,
+                                                              bool geoloc = false) {
+    // Read the patterns/targets    std::vector<float> lat = readCice(fileName, "ULAT");
```
Comment (CoryMartin-NOAA) [inline IceEmul/IceEmul.h]: why no lat?

File: test/daml_ice_emul.yaml [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+prediction:
+  # cice history: "/home/gvernier/data/20210715/00/gdas.t00z.icef009.nc"
+  cice history: "/home/gvernier/data/20210710/00/gdas.t00z.icef003.nc"
+  output filename: "gdas.t00z.icef009.ffnn.nc.20210710"
+
+#training:
+#  epochs: 200
+#  cice history: "/home/gvernier/data/20210715/00/gdas.t00z.icef009.nc"
+#  model output: "model.pt"
```
Comment (CoryMartin-NOAA) [inline test/daml_ice_emul.yaml]: no EOL

File: CMakeLists.txt (hunk 1 above)
Comment (guillaumevernieres) [reply CMakeLists.txt]: oOps. Bad copy/paste!

File: IceEmul/IceEmul.h (hunk 5 above)
Comment (guillaumevernieres) [reply IceEmul/IceEmul.h]: ouch ... good catch!

**Review by CoryMartin-NOAA — APPROVED** [review]

**Pull request merged**